File: breezy/treebuilder.py
# Copyright (C) 2006, 2008, 2009, 2010 Canonical Ltd
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

"""TreeBuilder helper class.

TreeBuilders are used to build trees of various shapes or properties. This
can be extremely useful in testing for instance.
"""

from . import errors


class AlreadyBuilding(errors.BzrError):

    _fmt = "The tree builder is already building a tree."


class NotBuilding(errors.BzrError):

    _fmt = "Not currently building a tree."


class TreeBuilder(object):
    """A TreeBuilder allows the creation of specific content in one tree at a
    time.
    """

    def __init__(self):
        """Construct a TreeBuilder."""
        self._tree = None
        self._root_done = False

    def build(self, recipe):
        """Build recipe into the current tree.

        :param recipe: A sequence of paths. For each path, the corresponding
            path in the current tree is created and added. If the path ends in
            '/' then a directory is added, otherwise a regular file is added.
        """
        self._ensure_building()
        if not self._root_done:
            self._tree.add('', 'directory', ids=b'root-id')
            self._root_done = True
        for name in recipe:
            if name.endswith('/'):
                self._tree.mkdir(name[:-1])
            else:
                end = b'\n'
                content = b"contents of %s%s" % (name.encode('utf-8'), end)
                self._tree.add(name, 'file')
                self._tree.put_file_bytes_non_atomic(name, content)

    def _ensure_building(self):
        """Raise NotBuilding if there is no current tree being built."""
        if self._tree is None:
            raise NotBuilding

    def finish_tree(self):
        """Finish building the current tree."""
        self._ensure_building()
        tree = self._tree
        self._tree = None
        tree.unlock()

    def start_tree(self, tree):
        """Start building on tree.

        :param tree: A tree to start building on. It must provide the
            MutableTree interface.
        """
        if self._tree is not None:
            raise AlreadyBuilding
        self._tree = tree
        self._tree.lock_tree_write()

File: src/treebuilder.rs
//! TreeBuilder helper class.
//!
//! TreeBuilders are used to build trees of various shapes or properties. This
//! can be extremely useful in testing for instance.

use crate::tree::{Error, MutableTree};

/// A TreeBuilder allows the creation of specific content in one tree at a time.
pub struct TreeBuilder<T: MutableTree> {
    tree: Option<T>,
    root_done: bool,
}

impl<T: MutableTree> TreeBuilder<T> {
    /// Creates a new empty `TreeBuilder`.
    ///
    /// The builder starts with no tree and no root directory.
    pub fn new() -> TreeBuilder<T> {
        TreeBuilder {
            tree: None,
            root_done: false,
        }
    }

    /// Starts building a new tree.
    ///
    /// This method locks the tree for writing and prepares it for building.
    /// It must be called before any other operations on the tree.
    ///
    /// # Arguments
    ///
    /// * `tree` - The tree to start building with.
    ///
    /// # Panics
    ///
    /// Panics if the builder already has a tree.
    pub fn start_tree(&mut self, mut tree: T) {
        if self.tree.is_some() {
            panic!("TreeBuilder already has a tree");
        }
        tree.lock_tree_write().expect("lock_tree_write");
        self.tree = Some(tree);
        self.root_done = false;
    }

    /// Finishes building the tree and returns it.
    ///
    /// This method unlocks the tree and returns it, allowing it to be used
    /// normally again.
    ///
    /// # Returns
    ///
    /// The completed tree.
    ///
    /// # Panics
    ///
    /// Panics if the builder does not have a tree.
    pub fn finish_tree(&mut self) -> T {
        let mut tree = if let Some(tree) = self.tree.take() {
            tree
        } else {
            panic!("TreeBuilder does not have a tree");
        };
        tree.unlock().expect("unlock");
        tree
    }

    /// Build recipe into the current tree.
    ///
    /// Args:
    ///   recipe: A sequence of paths. For each path, the corresponding
    ///     path in the current tree is created and added. If the path ends in
    ///     '/' then a directory is added, otherwise a regular file is added.
    pub fn build(&mut self, recipe: &[&str]) -> Result<(), Error> {
        let tree = if let Some(tree) = self.tree.as_mut() {
            tree
        } else {
            panic!("TreeBuilder does not have a tree");
        };
        if !self.root_done {
            tree.add(&[""])?;
            self.root_done = true;
        }
        for name in recipe {
            if name.ends_with('/') {
                tree.mkdir(name.trim_end_matches('/'))?;
            } else {
                let content = format!("contents of {}\n", name);
                tree.put_file_bytes_non_atomic(name, content.as_bytes())?;
                tree.add(&[name])?;
            }
        }
        Ok(())
    }
}

impl<T: MutableTree> Default for TreeBuilder<T> {
    fn default() -> Self {
        Self::new()
    }
}
